simulators/ethereum/engine/enginetests.go

```diff
@@ -0,0 +1,837 @@
+	if resp.Status != "SYNCING" {
+		t.Fatalf("FAIL (%v): Response on forkchoiceUpdated with unknown HeadBlockHash is not SYNCING: %v", t.TestName, resp)
+	}
+
+	// Test again using PayloadAttributes, should also return SYNCING and no PayloadID
+	payloadAttr := catalyst.PayloadAttributesV1{
+		Timestamp:             t.CLMock.LatestExecutedPayload.Timestamp + 1,
+		Random:                common.Hash{},
+		SuggestedFeeRecipient: common.Address{},
+	}
+	resp, err = t.Engine.EngineForkchoiceUpdatedV1(t.Engine.Ctx(), &forkchoiceStateUnknownHeadHash, &payloadAttr)
+	if err != nil {
+		t.Fatalf("FAIL (%v): Error on forkchoiceUpdated with unknown HeadBlockHash: %v, %v", t.TestName, err)
+	}
+	if resp.Status != "SYNCING" {
+		t.Fatalf("FAIL (%v): Response on forkchoiceUpdated with unknown HeadBlockHash is not SYNCING: %v, %v", t.TestName, resp)
+	}
+	if resp.PayloadID != nil {
+		t.Fatalf("FAIL (%v): Response on forkchoiceUpdated with unknown HeadBlockHash contains PayloadID: %v, %v", t.TestName, resp)
+	}
+
+}
+
+// Verify that a forkchoiceUpdated fails on hash being set to a pre-TTD block after PoS change
+func preTTDFinalizedBlockHash(t *TestEnv) {
+	// Wait until TTD is reached by this client
+	if !t.WaitForPoSSync() {
+		t.Fatalf("FAIL (%v): Timeout on PoS sync", t.TestName)
+	}
+
+	// Wait for ExecutePayload
+	if closed := t.CLMock.OnFinalizedBlockForkchoiceUpdate.Wait(); closed {
+		t.Fatalf("FAIL (%v): CLMocker stopped producing blocks", t.TestName)
+	}
+	defer t.CLMock.OnFinalizedBlockForkchoiceUpdate.Yield()
+
+	// Send the Genesis block as forkchoice
+	gblock := loadGenesis()
+	forkchoiceStateGenesisHash := catalyst.ForkchoiceStateV1{
+		HeadBlockHash:      gblock.Hash(),
+		SafeBlockHash:      gblock.Hash(),
+		FinalizedBlockHash: gblock.Hash(),
+	}
+	resp, err := t.Engine.EngineForkchoiceUpdatedV1(t.Engine.Ctx(), &forkchoiceStateGenesisHash, nil)
+
+	/* TBD: Behavior on this edge-case is undecided, as behavior of the Execution client
+	 		if not defined on re-orgs to a point before the latest finalized block.
+
+	if err == nil {
+		t.Fatalf("FAIL (%v): No error forkchoiceUpdated with genesis: %v, %v", t.TestName, err, resp)
+	}
+	*/
+
+	resp, err = t.Engine.EngineForkchoiceUpdatedV1(t.Engine.Ctx(), &t.CLMock.LatestForkchoice, nil)
+	if err != nil {
+		t.Fatalf("FAIL (%v): Error on forkchoiceUpdated with unknown FinalizedBlockHash: %v, %v", t.TestName, err)
+	}
+	if resp.Status != "SUCCESS" {
+		t.Fatalf("FAIL (%v): Response on forkchoiceUpdated with genesis is not SUCCESS: %v, %v", t.TestName, resp)
+	}
+}
+
+// Corrupt the hash of a valid payload, client should reject the payload
+func badHashOnExecPayload(t *TestEnv) {
+	// Wait until TTD is reached by this client
+	if !t.WaitForPoSSync() {
+		t.Fatalf("FAIL (%v): Timeout on PoS sync", t.TestName)
+	}
+
+	// Wait for GetPayload
+	if closed := t.CLMock.OnGetPayload.Wait(); closed {
+		t.Fatalf("FAIL (%v): CLMocker stopped producing blocks", t.TestName)
+	}
+	defer t.CLMock.OnGetPayload.Yield()
```

Comment:
The wait/yield looks pretty nice enough, but it's not what I'd call go-idiomatic - more pythonic maybe? 

We typically do it something like this: 
```suggestion
        select{ 
            case payload := < t.CLMock.payloadCh:
               // got payload ok
             case time.After(1 * time.Second) : 
               // time-out
               	t.Fatalf("FAIL (%v): CLMocker timed out", t.TestName)
              case shutdownCh:
               // shutdown channel closed
               	t.Fatalf("FAIL (%v): CLMocker stopped producing blocks", t.TestName)
       }
```
Having a `select` means we can act on the chan we're interested in (in this case `payloadCh`), but we can also have other cases, to handle timeouts (typically good to have in case there's  some deadlock somewhere) and/or if the whole thing is shutting down. 
Then for the payloadCh, you'd only ever deliver payloads over it. If/when you want to shutdown, you just close the `shutdownCh`. Any/all listeners currently on such a `select` will _all_ react to the `shutdownCh` close, and they can all exit nicely.